source/common/router/router.cc

```diff
@@ -329,10 +346,19 @@ void Filter::chargeUpstreamCode(Http::Code code,
 }
 
 Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers, bool end_stream) {
-  // Do a common header check. We make sure that all outgoing requests have all HTTP/2 headers.
-  // These get stripped by HTTP/1 codec where applicable.
-  ASSERT(headers.Method());
-  ASSERT(headers.Host());
+  // Do a common header check for minimum constraints. We make sure that all outgoing requests have
+  // all HTTP/2 headers.
+  const auto header_status = FilterUtility::checkHeaderMap(headers);
+  if (!header_status.ok()) {
+    callbacks_->streamInfo().setResponseFlag(StreamInfo::ResponseFlag::DownstreamProtocolError);
+    ENVOY_LOG_MISC(info, "HEADER STATUS MESSAGE {}", header_status.message());
+    const std::string body = fmt::format("missing required header: {}", header_status.message());
+    const std::string details = absl::StrCat(StreamInfo::ResponseCodeDetails::get().DirectResponse,
```

Comment:
Thanks for fixing this!  5xx SGTM.
Please don't reuse details though - the goal is for each to be used once so you can pinpoint the area of code triggering the response.
When you add a new one (FilterRemovedRequiredHeaders or whatever) you'll need to reflect the addition in docs/ as well

Reply:
Ah good point about the detail reuse -- I'll make a unique one for this type of error that indicates some kind of filter misbehavior!